src/libopensc/pkcs15-starcos-esign.c

```diff
@@ -0,0 +1,364 @@
+			prkey_obj.flags = containers[i].prdata->obj_flags;
+			if (containers[i].prdata->auth_id) {
+				sc_pkcs15_format_id(containers[i].prdata->auth_id, &prkey_obj.auth_id);
+			}
+
+			r = sc_pkcs15emu_add_rsa_prkey(p15card, &prkey_obj, &prkey_info);
+			LOG_TEST_RET(card->ctx, r, "Failed to add RSA prkey");
+		}
+
+		containers_added++;
+	}
+
+	if (containers_added == 0) {
+		r = SC_ERROR_INVALID_CARD;
+	} else {
+		r = SC_SUCCESS;
+	}
+
+	LOG_FUNC_RETURN(card->ctx, r);
+}
+
+/*
+ * adds the PKCS15 objects of the ESIGN application. The app may contain
+ * 1) Authentication container
+ *  - 2048-bit RSA key
+ *  - CH certificate
+ * 2) Encryption container
+ *  - 2048-bit RSA key
+ *  - CH certificate
+ * 3) Authentication Issuer container
+ *  - issuer certificate
+ * 3) Encryption Issuer container
+ *  - issuer certificate
+ * Depending on the card profile, some containers may be missing.
+ * Both RSA keys are protected with the UserPIN. The app may have a PUK, not
+ * supported by this emulator.
+ * 
+ * The issuer certificates are not included by default, define ADD_ISSUER_CONTAINERS
+ * to enable them.
+ */
+static int
+starcos_add_esign_app(sc_pkcs15_card_t *p15card)
+{
+	static cdata auth_cert = {"C.CH.AUT", 0, "3F00060843F1", "1", 0};
+	static cdata encr_cert = {"C.CH.ENC", 0, "3F0006084301", "2", 0};
+#ifdef ADD_ISSUER_CONTAINERS
+	const cdata auth_root_cert = { "C.RootCA_Auth", 1, "3F00060843F0", "3", 0 };
+	const cdata encr_root_cert = { "C.RootCA_Enc", 1, "3F0006084300", "4", 0 };
+#endif
+
+	static prdata auth_key = {"1", "PrK.CH.AUT", 2048, USAGE_AUT, "3F000608", 0x81, "1", SC_PKCS15_CO_FLAG_PRIVATE};
+	static prdata encr_key = {"2", "PrK.CH.ENC", 2048, USAGE_KE, "3F000608", 0x83, "1", SC_PKCS15_CO_FLAG_PRIVATE};
+
+	static pindata auth_pin = {"1", "UserPIN", "3F00", 0x01, SC_PKCS15_PIN_TYPE_UTF8, 16, 6, 0,
+			USER_PIN, -1, 3, 0x00, SC_PKCS15_CO_FLAG_MODIFIABLE | SC_PKCS15_CO_FLAG_PRIVATE};
+
+	static pindata auth_pin_v35 = {"1", "UserPIN", "3F00", 0x06, SC_PKCS15_PIN_TYPE_UTF8, 16, 6, 0,
+			USER_PIN, -1, 3, 0x00, SC_PKCS15_CO_FLAG_MODIFIABLE | SC_PKCS15_CO_FLAG_PRIVATE};
+
+	ppindata auth = (p15card->card->type == SC_CARD_TYPE_STARCOS_V3_5) ? &auth_pin_v35 : &auth_pin;
+
+	const container containers[] = {
+			{"1", &auth_cert, auth, &auth_key},
+			{"2", &encr_cert, auth, &encr_key},
+#ifdef ADD_ISSUER_CONTAINERS
+			{ "3", &auth_root_cert, 0, 0 },
+			{ "4", &encr_root_cert, 0, 0 },
+#endif
+	};
+
+	return add_app(p15card, containers, sizeof(containers) / sizeof(container));
+}
+
+static int
+starcos_esign_init(sc_pkcs15_card_t *p15card, struct sc_aid *aid)
+{
+	sc_card_t *card = p15card->card;
+	sc_context_t *ctx = card->ctx;
+	const char *label = name_Card;
+	int r;
+	int apps_added = 0;
+
+	SC_FUNC_CALLED(ctx, SC_LOG_DEBUG_NORMAL);
+
+	if (card->type != SC_CARD_TYPE_STARCOS_V3_4 && card->type != SC_CARD_TYPE_STARCOS_V3_5) {
+		return SC_ERROR_WRONG_CARD;
+	}
```

Comment:
By nature, STARCOS is a chip OS that can be configured with a number of profiles. If someone had the time to implement this correctly, then parsing of EF_KEYD, EF_CERT, ... should be done. For sake of simplicity and due to encoding problems during card initialization, we often take shortcuts by providing a hard coded profile as done in [pkcs15-starcos-esign.c](https://github.com/OpenSC/OpenSC/pull/2544/files#diff-b8cf394129d75ae702ab117c61b6fd7d86761bf056b340844ac402dadf546057). However, I think you want to avoid matching generic STARCOS 3.5 cards (which may not exactly have the profile you are assuming here...)

Typically, OpenSC driver developers use a dedicated card type, say `SC_CARD_TYPE_STARCOS_V3_5_VDA`, which not only matches the STARCOS version, but also the specific ATRs of your cards. This way, you can use your emulator for your cards, but for other STARCOS 3.5 cards, it would still be possible to parse the default ISO7815-15 structure to search for keys, certificates and apps.

Reply:
Unfortunately the card profile has no unique ATR (at least for the StarCOS 3.4), I have a D-Trust card which responds with the same ATR, but with a completely different profile. That card has a correct EF_DIR and a proper #PKCS15 structure on card, and works directly with OpenSC. The cards with the eSign profile have several issues, some lacks the ISO #PKCS15 directory and some have a broken version. This emulator tries to fix these issues.
Anyway, I will double check if I can assign these cards a uniqe card type.

Reply:
I have double checked the ATRs of this profile.
StarCOS 3.5 eSign card seems to have a unique ATR, so a new card type SC_CARD_TYPE_STARCOS_V3_5_ESIGN could be defined. Unfortunately, in case of the v3.4 card, the profile uses a generic ATR, so AID checking is needed to determine if the card is SC_CARD_TYPE_STARCOS_V3_4_ESIGN.
I set these card types as emulation only.